src/tasks/block.rs

```diff
@@ -109,38 +110,69 @@ impl InProgressBlock {
     pub config: BuilderConfig,
+    pub tx_poller: TxPoller,
+    pub bundle_poller: BundlePoller,
 }
 
 impl BlockBuilder {
     // create a new block builder with the given config.
-    pub fn new(config: &BuilderConfig) -> Self {
+    pub fn new(config: &BuilderConfig, authenticator: Authenticator) -> Self {
         Self {
             config: config.clone(),
             incoming_transactions_buffer: config.incoming_transactions_buffer,
+            tx_poller: TxPoller::new(config),
+            bundle_poller: BundlePoller::new(config, authenticator),
         }
     }
 
-    /// Spawn the block builder task, returning the inbound channel to it, and
-    /// a handle to the running task.
-    pub fn spawn(
-        self,
-        outbound: mpsc::UnboundedSender<InProgressBlock>,
-    ) -> (mpsc::UnboundedSender<TxEnvelope>, mpsc::UnboundedSender<Bundle>, JoinHandle<()>) {
-        let mut in_progress = InProgressBlock::default();
+    async fn get_transactions(&mut self, in_progress: &mut InProgressBlock) {
+        let txns = self.tx_poller.check_tx_cache().await;
+        match txns {
+            Ok(txns) => {
+                for txn in txns.into_iter() {
+                    in_progress.ingest_tx(&txn);
+                }
+            }
+            Err(e) => {
+                tracing::error!("error polling transactions: {:?}", e);
+            }
+        }
+        self.tx_poller.evict();
+    }
 
-        let (tx_sender, mut tx_inbound) = mpsc::unbounded_channel();
-        let (bundle_sender, mut bundle_inbound) = mpsc::unbounded_channel();
+    async fn get_bundles(&mut self, in_progress: &mut InProgressBlock) {
+        let bundles = self.bundle_poller.check_bundle_cache().await;
+        match bundles {
+            Ok(bundles) => {
+                for bundle in bundles {
+                    in_progress.ingest_bundle(bundle);
+                }
+            }
+            Err(e) => {
+                tracing::error!("error polling bundles: {:?}", e);
+            }
+        }
+        self.bundle_poller.evict();
+    }
 
+    /// Spawn the block builder task, returning the inbound channel to it, and
+    /// a handle to the running task.
+    pub fn spawn(mut self, outbound: mpsc::UnboundedSender<InProgressBlock>) -> JoinHandle<()> {
         let mut sleep =
             Box::pin(tokio::time::sleep(Duration::from_secs(self.incoming_transactions_buffer)));
 
-        let handle = tokio::spawn(
+        tokio::spawn(
             async move {
                 loop {
 
                     select! {
                         biased;
                         _ = &mut sleep => {
+                            // Build a block
+                            let mut in_progress = InProgressBlock::default();
+                            self.get_transactions(&mut in_progress).await;
+                            self.get_bundles(&mut in_progress).await;
+
                             if !in_progress.is_empty() {
```

Comment:
So, as far as I understand now, instead of spawning a loop that awaits 3 different futures concurrently, we now have a loop that:
1. gets all txs
2. gets all bundles
3. builds the block
4. sends it
5. sleeps for X seconds

In this case we don't need the `select!` at all anymore. We just need to loop like we already are and go through the steps.

Reply:
good shout https://github.com/init4tech/builder/pull/22/commits/ada0cd28cb6ab1585cb7e978809448f7a6980b07